packages/@expo/cli/src/utils/codesigning.ts

```diff
@@ -0,0 +1,350 @@
+  return null;
+}
+
+/**
+ * Get the certificate configured for expo-updates for this project.
+ */
+async function getProjectCodeSigningCertificateAsync(
+  exp: ExpoConfig,
+  privateKeyPath: string | undefined,
+  expectedKeyId: string,
+  expectedAlg: string | null
+): Promise<CodeSigningInfo | null> {
+  const codeSigningCertificatePath = exp.updates?.codeSigningCertificate;
+  if (!codeSigningCertificatePath) {
+    return null;
+  }
+
+  if (!privateKeyPath) {
+    privateKeyPath = path.join(path.dirname(codeSigningCertificatePath), 'private-key.pem');
+  }
+
+  const codeSigningMetadata = exp.updates?.codeSigningMetadata;
+  if (!codeSigningMetadata) {
+    throw new Error(
+      'Must specify "codeSigningMetadata" under the "updates" field of your app config file to use EAS code signing'
+    );
+  }
+
+  const { alg, keyid } = codeSigningMetadata;
+  if (!alg || !keyid) {
+    throw new Error(
+      'Must specify "keyid" and "alg" in the "codeSigningMetadata" field under the "updates" field of your app config file to use EAS code signing'
+    );
+  }
+
+  if (expectedKeyId !== keyid) {
+    throw new Error(`keyid mismatch: client=${expectedKeyId}, project=${keyid}`);
+  }
+
+  if (expectedAlg && expectedAlg !== alg) {
+    throw new Error(`alg mismatch: client=${expectedAlg}, project=${alg}`);
+  }
+
+  const { privateKeyPEM, certificatePEM } =
+    await getProjectPrivateKeyAndCertificateFromFilePathsAsync({
+      codeSigningCertificatePath,
+      privateKeyPath,
+    });
+
+  return {
+    privateKey: privateKeyPEM,
+    certificateForPrivateKey: certificatePEM,
+    certificateChainForResponse: [],
+  };
+}
+
+async function readFileWithErrorAsync(path: string, errorMessage: string): Promise<string> {
+  try {
+    return await fs.readFile(path, 'utf8');
+  } catch {
+    throw new Error(errorMessage);
+  }
+}
+
+async function getProjectPrivateKeyAndCertificateFromFilePathsAsync({
+  codeSigningCertificatePath,
+  privateKeyPath,
+}: {
+  codeSigningCertificatePath: string;
+  privateKeyPath: string;
+}): Promise<{ privateKeyPEM: string; certificatePEM: string }> {
+  const [codeSigningCertificatePEM, privateKeyPEM] = await Promise.all([
+    readFileWithErrorAsync(
+      codeSigningCertificatePath,
+      `Code signing certificate cannot be read from path: ${codeSigningCertificatePath}`
+    ),
+    readFileWithErrorAsync(
+      privateKeyPath,
+      `Code signing private key cannot be read from path: ${privateKeyPath}`
+    ),
+  ]);
+
+  const privateKey = convertPrivateKeyPEMToPrivateKey(privateKeyPEM);
+  const certificate = convertCertificatePEMToCertificate(codeSigningCertificatePEM);
+  validateSelfSignedCertificate(certificate, {
+    publicKey: certificate.publicKey as PKI.rsa.PublicKey,
+    privateKey,
+  });
+
+  return { privateKeyPEM, certificatePEM: codeSigningCertificatePEM };
+}
+
+function validateStoredDevelopmentExpoRootCertificateCodeSigningInfo(
```

Comment:
Any additional comments would be appreciated, I have no particular formatting in mind but I think anything would be useful.